File: go/Q24_closest_numbers/solution3.go
package main

func findClosestElements3(list []int, k int, number int) []int {
	left := 0
	right := len(list) - k

	for left < right {
		middle := (left + right) / 2

		if number-list[middle] > list[middle+k]-number {
			left = middle + 1
		} else {
			right = middle
		}
	}

	return list[left : left+k]
}

File: ruby/Q28_rain_trap/solution3.rb
def trap(heights)
  total = 0
  left = 0
  right = heights.length() - 1
  left_max = right_max = 0

  while left < right
    if heights[left] < heights[right]
      if heights[left] >= left_max
        left_max = heights[left]
      else
        total += left_max - heights[left]
      end
      left += 1
    else
      if heights[right] >= right_max
        right_max = heights[right]
      else
        total += right_max - heights[right]
      end
      right -= 1
    end
  end

  return total
end
